text/0026-transform_application.md

```diff
@@ -0,0 +1,107 @@
+
+At the same time, we have never had a full explanation for how transforms and FSI are supposed to work together, and there's lots of bugs now in how dry-run and transform interact with FSI. We're at a point where we need to re-evaluate how these pieces work together.
+
+Let's go back to the very beginning.
+
+## Manual Save
+
+In the early days, it was decided that the right way to think about a `save` operation on a dataset that already existed was as a "patch application". Let's say I have a dataset named "dustomp/my_ds" that looks like this:
+
+![an example dataset](https://github.com/qri-io/rfcs/blob/rfc-xform/img/dataset_01.png)
+
+Then I run this command to create a second version:
+
+```
+  qri save --file new_meta.json me/my_ds
+```
+
+What this means is "take my previous version, and *patch it* by updating the meta component with the contents of this file". Everything else in the dataset stays the same. In this way, a manual save operation acts as a state transition between versions by way of patching the previous to arrive at the next.
+
+![diagram of a patch application](https://github.com/qri-io/rfcs/blob/rfc-xform/img/dataset_02_patch_application.png)
+
+## Transforms
+
+When transform scripts were introduced, we needed a similar way to describe the semantics of how they operate. A major question early on was whether a transform living with a dataset should always be reapplied on each save operation, or if it should only be used once, on the commit at which it was added.
+
+We went back and forth on this question, and ultimately arrived at the following functionality: a transform should only be executed on the commit where it gets added, and not every commit in the future, because the resulting dataset may not make sense to use as input to the same transform (think about a transform that adds or removes a column). However, if a user wanted to reapply the same transform, they could opt-in to that functionality by using the `--recall` flag, which would retrieve a recently rerun transform to run again.
+
+This transform changes the shape of the data, so it doesn't make sense to apply it again:
+![transform that changes shape](https://github.com/qri-io/rfcs/blob/rfc-xform/img/dataset_03_xform_diff_shape.png)
+
+This transform keeps the same shape of the data; running it a second time makes sense:
+![transform that maintains shape](https://github.com/qri-io/rfcs/blob/rfc-xform/img/dataset_04_xform_same_shape.png)
+
+Similar to a manual save, a transform script represents a state transition, with a reproducible method to execute that state transition, and a mechanism called "recall" to clone that state transition when applicable.
+
+A crucial fact about this description is that a manual save and a transform cannot both happen at the same time. They are both state transitions, and since we define a linear history of commits, it's illegal to have two transitions at once. Neither forking nor merging are allowed:
```

Comment:
Hrm... you _can_ mix manual & scripted transforms in the same commit. both just can't modify the same path.